### PR TITLE
Clean up named tensor `propagate_names` API

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29322 named tensor support for torch.equal
* #29307 Make TensorName::unifyFromRight in-place for efficiency
* **#29239 Clean up named tensor `propagate_names` API**

There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

Differential Revision: [D18350090](https://our.internmc.facebook.com/intern/diff/D18350090)

### CODE REVIEW

File: aten/src/ATen/NamedTensorUtils.cpp [hunk 1]
```diff
@@ -128,51 +127,41 @@ static void assert_names_equal(DimnameList a, DimnameList b) {
...
-  propagate_names(result.unsafeGetTensorImpl(), names);
+Tensor& propagate_names(Tensor& result, DimnameList names, bool validate_names) {
+  propagate_names(result.unsafeGetTensorImpl(), names, validate_names);
+  return result;
 }
 
-void propagate_names(Tensor& result, std::vector<Dimname>&& names, bool validate_names) {
-  propagate_names(result.unsafeGetTensorImpl(), std::move(names), validate_names);
+TensorImpl* propagate_names(TensorImpl* result, DimnameList names, bool validate_names) {
+  if (result->sizes().size() > 0) {
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.cpp]: I think `result->dim()` is clearer if I'm correctly understanding this

File: aten/src/ATen/NamedTensorUtils.cpp [hunk 2]
```diff
@@ -461,20 +453,20 @@ optional<std::vector<Dimname>> compute_cat_outnames(TensorList tensors) {
   return result;
 }
 
-optional<std::vector<Dimname>> compute_matmul_outnames(
+std::vector<Dimname> compute_matmul_outnames(
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.cpp]: I like this change. I think I mentioned before I generally prefer avoiding optional/nullable lists wherever possible.

File: aten/src/ATen/NamedTensorUtils.h [hunk 3]
```diff
@@ -43,26 +43,97 @@ unify_from_right(DimnameList names, DimnameList other, const char* action = "bro
...
-// 2) If result has names, then `names` must be equal to result.names
-void propagate_names(Tensor& result, optional<DimnameList> names);
-void propagate_names(Tensor& result, std::vector<Dimname>&& names, bool validate_names);
-CAFFE2_API void propagate_names(Tensor& result, optional<std::vector<Dimname>>&& maybe_names, bool validate_names);
-void propagate_names(TensorImpl* result, optional<DimnameList> names);
-void propagate_names(TensorImpl* result, std::vector<Dimname>&& names, bool validate_names);
-void propagate_names(TensorImpl* result, optional<std::vector<Dimname>>&& maybe_names, bool validate_names);
+// [NOTE] Writing name inference rules
+//
+// Operators that support named tensors generally look like the following:
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.h]: I assume that in some cases ops that just call other ops could support named tensors without doing anything?

File: aten/src/ATen/NamedTensorUtils.h [hunk 4]
```diff
@@ -43,26 +43,97 @@ unify_from_right(DimnameList names, DimnameList other, const char* action = "bro
...
+// - non-empty outnames.
+//
+// propagate_names_if_nonempty propagates the outnames if they exist to the result
+// tensors.
+//
+// The {} case is an optimization; if the user does not use named tensors they
+// pay no perf cost for it.
+
+// Propagates `names` to `result` if `names` is not empty.
+// `names` can be empty; see [NOTE] Writing name inference rules
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.h]: this `see [NOTE]` seems recursive

**Review by nairbv — APPROVED** [review]

File: aten/src/ATen/NamedTensorUtils.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: I like the type safety that optional gives but I got too many headaches dealing with `optional<ArrayRef>` and `optional<vector>` so I decided to kill it here.

File: aten/src/ATen/NamedTensorUtils.h (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.h]: Yes. I'll include that in

File: aten/src/ATen/NamedTensorUtils.h (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.h]: There's a space between the [NOTE] and this block that mentions the note. This block is supposed to be documentation for `propagate_names_if_nonempty`.

Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.h]: I'll add an extra space to make it clearer

File: aten/src/ATen/NamedTensorUtils.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: Yes, thanks for catching that

File: aten/src/ATen/NamedTensorUtils.h (hunk 4 above)
Comment (nairbv) [reply aten/src/ATen/NamedTensorUtils.h]: ah I see. some other ideas: you could start/end the note with a line of `////////////`'s or move the note above the start of the namespace, or use /**  */'s to better distinguish it from comments on the specific function below.

**Pull request closed without merging**